aflplusplus: 4.10c -> 4.20c

Sigmanificient opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Sigmanificient committed:
aflplusplus: 4.10c -> 4.20c

Mindavi reviewed (dismissed):
Hum, that's not gonna work :). Gonna need at least an update of the hash as well. And upstream has updated their customized qemu, so I prefer to keep in sync with that too.

I started making the update but haven't found time to finalise it yet, would be great if you could have a look 🙂.

Sigmanificient commented:
> Hum, that's not gonna work :). Gonna need at least an update of the hash as well. And upstream has updated their customized qemu, so I prefer to keep in sync with that too.
> 
> I started making the update but haven't found time to finalise it yet, would be great if you could have a look 🙂.

Oh, I  completely forgot to update that hash, not sure why it built!

Mindavi commented:
Qemu should still be updated to the latest version upstream uses.

Sigmanificient commented:
> > Oh, I completely forgot to update that hash, not sure why it built!
> 
> Since the hash was the same, nix reused the old version's source. There are two things you could do to prevent that sort of issue:
> 
>     * add a test with [`testVersion`](https://nixos.org/manual/nixpkgs/unstable/#tester-testVersion) ;
> 
>     * use an [`updateScript`](https://nixos.org/manual/nixpkgs/unstable/#var-passthru-updateScript) instead of updating manually.
> 
> 
> @Sigmanificient, would you want to do that yourself, or should I do it prior to merging?

May you go ahead, I'm not fully sure how you want to do it

SuperSandro2000 commented:
Fails to build:


```
make: Leaving directory '/build/source'
[*] 12 test cases completed.
[-] not all test cases were executed
[!] failure in tests :-(
```

doronbehar commented:
@Sigmanificient would you like to keep working on this? I see 4.21c	was released 3 days ago:
https://github.com/AFLplusplus/AFLplusplus/releases

Sigmanificient commented:
> @Sigmanificient would you like to keep working on this? I see 4.21c was released 3 days ago: [AFLplusplus/AFLplusplus/releases](https://github.com/AFLplusplus/AFLplusplus/releases)

```

+++ Testing aborted programmatically +++
[!] Time limit was reached
[*] Statistics: 0 new corpus items found, 0.11% coverage achieved, 0 crashes saved, 0 timeouts saved, total runtime 0 days, 0 hrs, 0 min, 7 sec
[+] We're done here. Have a nice day!

CUT------------------------------------------------------------------CUT
[!] afl-fuzz is not working correctly with qemu_mode cmplog
[*] running afl-fuzz for persistent qemu_mode, this will take appro
```
Tried to build with `4.21c` no sure what to do about this

doronbehar commented:
That is indeed weird. If you don't find the cause for this failure we can disable the tests in the worse case, with a comment explaining this situation. If you want to continue debug this, I'd try to do so in a `nix develop` environment in a git checkout.

Sigmanificient commented:
I clearly underestimated the difficulty of this package when i opened that pr quite some time ago  :sweat_smile:
Feel like someone else would do a much better job, especially since i dont have much free time currently

doronbehar commented:
Yes it would have been nice if the maintainers would have responded.. If you are not sure whether you are breaking something with the update, and you haven't tested it properly, it's legitimate to let it go.

Mindavi commented:
I'll probably pick this up myself soon-ish, but I'm generally a bit conservative with updating this since the upstream build system is not very nixpkgs-friendly. But I'll definitely pick this whenever I do so, so thanks for that.

Mindavi committed:
aflplusplus: 4.20c: update qemuafl to match 4.20c

Sigmanificient commented:
This does build

Sigmanificient commented:
I thought it was all good,  but tests are still failling sadly :<

Mindavi commented:
I don't really understand, it seemed to work fine locally but maybe I mislooked... The unit tests are less interesting anyway though.

doronbehar commented:
You may want to try to rebase the PR branch on top of latest `master` to reproduce the failure.

Mindavi commented:
Hmm, even after a rebase on master:

```
aflplusplus> [  PASSED  ] 1 test(s).
aflplusplus> make: Leaving directory '/build/source'
aflplusplus> [*] 12 test cases completed.
aflplusplus> [-] not all test cases were executed
aflplusplus> [+] all tests were successful :-)
aflplusplus> installCheckPhase completed in 3 minutes 7 seconds
```

Let's have another attempt.

@ofborg build aflplusplus

doronbehar commented:
Hmm still fails...

Mindavi commented:
This build seems to have succeeded: https://github.com/NixOS/nixpkgs/pull/305773/checks?check_run_id=27428395866.

doronbehar commented:
> This build seems to have succeeded: [#305773 (checks)](https://github.com/NixOS/nixpkgs/pull/305773/checks?check_run_id=27428395866).

Right! What I saw was an old attempt probably. You edited your previous comment and I think ofborg didn't catch the change `eval` -> `build alfplusplus`... Never mind, it seems to be OK now.

doronbehar commented on pkgs/tools/security/aflplusplus/qemu.nix:
```diff
@@ -24,8 +24,8 @@ stdenv.mkDerivation {
     repo = "qemuafl";
-    rev = "a1321713c7502c152dd7527555e0f8a800d55225";
-    sha256 = "sha256-HLlOHqT2vrHjHyu4n83IzVzKv9ErinephLLev1E10nM=";
+    rev = "40033af00c4c5de172ed4fe60c21b9edbd2c189d";
```
In the next update of AFLplusplus, how would we be able to know what revision to use? Is this a submodule or something alike?

Sigmanificient replied:
I think this is this: https://github.com/AFLplusplus/AFLplusplus/commits/stable/qemu_mode/QEMUAFL_VERSION

Sigmanificient replied:
This commit https://github.com/AFLplusplus/AFLplusplus/commit/07bc202e0ad940e0cc7c8770f69ceb32ed851384 sets it to `40033af`

doronbehar replied:
> I think this is this: [AFLplusplus/AFLplusplus@`stable`/qemu_mode/QEMUAFL_VERSION (commits)](https://github.com/AFLplusplus/AFLplusplus/commits/stable/qemu_mode/QEMUAFL_VERSION)

@Sigmanificient this is a great link to put in a comment near that revision. Could you please do that? Besides that this PR can be merged IMO.

doronbehar reviewed (commented):
Looks good overall. Should be ready to merge once this small issue is fixed.

Sigmanificient commented:
@doronbehar I noticed a depreciated `replace` (substitueInPlace) while building, may I update them?

doronbehar commented:
> @doronbehar I noticed a depreciated `replace` (substitueInPlace) while building, may I modernize the derivation a bit?

Yea sure go ahead. I'd recommend do so in a separate commit, and before the update commit.

Sigmanificient committed:
aflplusplus: modernize

- Update depreciated `--replace` by `replace-fail`
- Replace `$(out)` with placeholder
- Change `sha` attributes to `hash`

Sigmanificient commented:
One of the changes is `sha` -> `hash`, so i keep the commit after to avoid the conflict

doronbehar reviewed (commented):
If you are attempting to modernize this, how about go all the way? :)

Use `nixfmt` (from package `nixfmt-rfc-style`), and use `finalAttrs` (function argument to `mkDerivation`). You can put each step in a dedicated PR.

Bonus: https://github.com/NixOS/nixpkgs/issues/208242 .

Sigmanificient commented:
Well tough that running `nixfmt` would ofuscate too much the diff

Sigmanificient commented:
Result of `nixpkgs-review pr 305773` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>aflplusplus</li>
    <li>libdislocator</li>
  </ul>
</details>

Sigmanificient committed:
afplusplus: format

Sigmanificient commented:
Applied formatting and fixed #208242 
I will not try to migrate it to by-name though :sweat_smile:

Mindavi reviewed (approved):
Let's see what ofborg thinks now. Thanks for modernizing this package description<3

Sigmanificient commented:
Result of `nixpkgs-review pr 305773` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>aflplusplus</li>
    <li>libdislocator</li>
  </ul>
</details>
